improvedigitaladserverjsclient.js

```diff
@@ -0,0 +1,230 @@
+      'errorCode': errorCode
+    };
+  };
+
+  this.createRequest = function (requestObject, requestParameters, extraRequestParameters) {
+    if (requestParameters.httpRequestType !== this.CONSTANTS.HTTP_REQUEST_TYPE.GET) {
+      return this.getErrorReturn(this.CONSTANTS.ERROR_CODES.BAD_HTTP_REQUEST_TYPE_PARAM);
+    }
+    if (!requestParameters.libVersion) {
+      return this.getErrorReturn(this.CONSTANTS.ERROR_CODES.LIB_VERSION_MISSING);
+    }
+
+    requestParameters.returnObjType = requestParameters.returnObjType || this.CONSTANTS.RETURN_OBJ_TYPE.DEFAULT;
+
+    var impressionObjects = [];
+    var impressionObject = void 0;
+    var counter = void 0;
+    if (utils.isArray(requestObject)) {
+      for (counter = 0; counter < requestObject.length; counter++) {
+        impressionObject = this.createImpressionObject(requestObject[counter]);
+        impressionObjects.push(impressionObject);
+      }
+    } else {
+      impressionObject = this.createImpressionObject(requestObject);
+      impressionObjects.push(impressionObject);
+    }
+
+    var returnIdMappings = true;
+    if (requestParameters.returnObjType === this.CONSTANTS.RETURN_OBJ_TYPE.PREBID) {
+      returnIdMappings = false;
+    }
+
+    var returnObject = {};
+    returnObject.requests = [];
+    if (returnIdMappings) {
+      returnObject.idMappings = [];
+    }
+    var errors = null;
+
+    var baseUrl = (requestParameters.secure === 1 ? 'https' : 'http') + '://' + this.CONSTANTS.AD_SERVER_BASE_URL + '/' + this.CONSTANTS.END_POINT + '?' + this.CONSTANTS.AD_SERVER_URL_PARAM;
+
+    var bidRequestObject = {
+      bid_request: this.createBasicBidRequestObject(requestParameters, extraRequestParameters)
+    };
+    for (counter = 0; counter < impressionObjects.length; counter++) {
+      impressionObject = impressionObjects[counter];
+
+      if (impressionObject.errorCode) {
+        errors = errors || [];
+        errors.push({
+          errorCode: impressionObject.errorCode,
+          adUnitId: impressionObject.adUnitId
+        });
+      } else {
+        if (returnIdMappings) {
+          returnObject.idMappings.push({
+            adUnitId: impressionObject.adUnitId,
+            id: impressionObject.impressionObject.id
+          });
+        }
+        bidRequestObject.bid_request.imp = bidRequestObject.bid_request.imp || [];
+        bidRequestObject.bid_request.imp.push(impressionObject.impressionObject);
+
+        var writeLongRequest = false;
+        var outputUri = baseUrl + encodeURIComponent(JSON.stringify(bidRequestObject));
+        if (outputUri.length > this.CONSTANTS.MAX_URL_LENGTH) {
+          writeLongRequest = true;
+          if (bidRequestObject.bid_request.imp.length > 1) {
+            // Pop the current request and process it again in the next iteration
+            bidRequestObject.bid_request.imp.pop();
+            if (returnIdMappings) {
+              returnObject.idMappings.pop();
+            }
+            counter--;
```

Comment:
Is there a chance that this causes a infinite loop? Can you please explain what this part (inside for loop) of the code is doing?

Reply:
Our client supports the submission of one impression per request or multiple impressions per request.  For technical reasons, we are not making use of single request in our adapters for now.  This code is only triggered in single request mode so there is no chance of an infinite loop.

Anyway, to explain the logic, if single request is turned on, the following happens:

1) Keep pushing impression objects into the request until it exceeds MAX_URL_LENGTH.
2) If it exceeds MAX_URL_LENGTH and there is more than 1 impression object, pop the last impression object from the request, decrement counter so that you can process it as part of another request.

Not that if a single impression request exceeds MAX_URL_LENGTH (which only applies to older browsers), it still gets sent since we don't apply this check for requests with just one impression in them.